### PR TITLE
Fix -classForCoder on bridged classes to respect mutability.

### LINKED ISSUE #902: Bridged class implementation of classForCoder always returns immutable class
This can be an issue if an app tries to encode/decode mutable types

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
* Add tests for all of the cases we support.
* We do not support archival of NSAttributedString (!)

Fixes #902.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1679)

### CODE REVIEW

File: tests/unittests/Foundation/NSAttributedStringTests.mm [hunk 1]
```diff
@@ -772,4 +772,19 @@ - (NSUInteger)length {
...
+    ASSERT_OBJCNE(nil, data);
+
+    NSMutableAttributedString* output = [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    ASSERT_OBJCNE(nil, output);
+
+    EXPECT_NO_THROW([output appendAttributedString:[[[NSMutableAttributedString alloc] initWithString:@" world"] autorelease]]);
+    EXPECT_OBJCEQ(@"hello world", [output string]);
+
+    EXPECT_OBJCNE(input, output);
+}
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSAttributedStringTests.mm]: This file has CRLF line endings.

File: tests/unittests/Foundation/NSCharacterSetTests.m [hunk 2]
```diff
@@ -222,4 +222,21 @@ void testCharacter(NSCharacterSet* charSet, NSString* testString) {
...
+    NSMutableCharacterSet* output = [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    ASSERT_OBJCNE(nil, output);
+
+    EXPECT_NO_THROW([output addCharactersInString:@"b"]);
+
+    // We cannot compare these for equality. Despite having different ranges (!) they are considered equal
+    // on the reference platform.
+    EXPECT_FALSE([input characterIsMember:'b']);
+    EXPECT_TRUE([output characterIsMember:'b']);
+}
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSCharacterSetTests.m]: This file has CRLF line endings.

File: Frameworks/Foundation/NSString.mm [hunk 3]
```diff
@@ -2021,7 +2021,13 @@ - (id)initWithCoder:(NSCoder*)coder {
  @Status Interoperable
 */
 - (void)encodeWithCoder:(NSCoder*)coder {
-    [coder encodeObject:self forKey:@"NS.string"];
+    if ([coder isKindOfClass:[NSKeyedArchiver class]]) {
```
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSString.mm]: I am not happy about having to do this.
However.

Without this escape hatch, our `NSKeyedArchiver` will encode `[NSMutableString stringWithString:@"x"]` as:

```
0: $null
1: {
    class: <2>
    NS.string: <1>
}
2: {
    $className: NSMutableString
    ...
}
```

(where the string is object **1** and `<x>` indicates a reference to object *x*)

Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSString.mm]: That is to say: NSKeyedArchiver will encode a string that refers to itself as its string.

**Comment by rajsesh-msft** [comment]
:shipit:

**Pull request merged**